core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,185 @@
+/// This component will upload Witness Inputs to the object store.
+/// This allows Witness Generator workflow (that needs only Basic Witness Generator Inputs)
+/// to be run only using the object store information, having no other external dependency.
+#[derive(Debug)]
+pub struct BasicWitnessInputProducer {
+    connection_pool: ConnectionPool,
+    validation_computational_gas_limit: u32,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl BasicWitnessInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool,
+        store_factory: &ObjectStoreFactory,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<Self> {
+        let validation_computational_gas_limit = u32::MAX;
+        Ok(BasicWitnessInputProducer {
+            connection_pool,
+            validation_computational_gas_limit,
+            object_store: store_factory.create_store().await.into(),
+            l2_chain_id,
+        })
+    }
+
+    fn process_job_impl(
+        rt_handle: Handle,
+        l1_batch_number: L1BatchNumber,
+        started_at: Instant,
+        connection_pool: ConnectionPool,
+        validation_computational_gas_limit: u32,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<WitnessBlockState> {
+        let mut connection = rt_handle.block_on(connection_pool.access_storage())?;
+
+        let miniblocks_execution_data = rt_handle.block_on(
+            connection
+                .transactions_dal()
+                .get_miniblocks_to_execute_for(MiniblockExecutionMode::L1Batch(l1_batch_number)),
+        );
+
+        let (mut vm, storage_view) = create_vm(
+            rt_handle.clone(),
+            l1_batch_number,
+            connection,
+            validation_computational_gas_limit,
+            l2_chain_id,
+        );
+
+        tracing::info!("Started execution of l1_batch: {l1_batch_number:?}");
+        for (index, miniblock_execution_data) in miniblocks_execution_data.iter().enumerate() {
+            tracing::debug!(
+                "Started execution of miniblock: {:?}",
+                miniblock_execution_data.number
+            );
+            for tx in &miniblock_execution_data.txs {
+                tracing::debug!("Started execution of tx: {tx:?}");
+                execute_tx(tx, &mut vm);
+                tracing::debug!("Finished execution of tx: {tx:?}");
+            }
+            if index + 1 < miniblocks_execution_data.len() {
+                vm.start_new_l2_block((&miniblocks_execution_data[index + 1]).into());
+            };
+
+            tracing::debug!(
+                "Finished execution of miniblock: {:?}",
+                miniblock_execution_data.number
+            );
+        }
+        vm.finish_batch();
+        tracing::info!("Finished execution of l1_batch: {l1_batch_number:?}");
+
+        metrics::histogram!(
+            "basic_witness_input_producer.input_producer_time",
+            started_at.elapsed(),
+        );
```

Comment:
Nit: Is it too much to ask to use `vise` metrics here. I know witness-related server components don't use them, but the remaining parts of the server do.